packages/baseten/src/baseten-provider.ts

```diff
@@ -0,0 +1,229 @@
+    fetch?: FetchFunction;
+  }
+
+  const getCommonModelConfig = (
+    modelType: string,
+    customURL?: string,
+  ): CommonModelConfig => ({
+    provider: `baseten.${modelType}`,
+    url: ({ path }) => {
+      // For embeddings with /sync URLs, we need to add /v1
+      if (modelType === 'embedding' && customURL?.includes('/sync') && !customURL?.includes('/sync/v1')) {
+        return `${customURL}/v1${path}`;
+      }
+      return `${customURL || baseURL}${path}`;
+    },
+    headers: getHeaders,
+    fetch: options.fetch,
+  });
+
+  const createChatModel = (modelId?: BasetenChatModelId) => {
+    // Use modelURL if provided, otherwise use default Model APIs
+    const customURL = options.modelURL;
+
+    if (customURL) {
+      // Check if this is a /sync/v1 endpoint (OpenAI-compatible) or /predict endpoint (custom)
+      const isOpenAICompatible = customURL.includes('/sync/v1');
+
+      if (isOpenAICompatible) {
+        // For /sync/v1 endpoints, use standard OpenAI-compatible format
+        return new OpenAICompatibleChatLanguageModel(modelId ?? 'placeholder', {
+          ...getCommonModelConfig('chat', customURL),
+          errorStructure: basetenErrorStructure,
+        });
+      } else if (customURL.includes('/predict')) {
+        throw new Error(
+          'Not supported. You must use a /sync/v1 endpoint for chat models.',
+        );
+      }
+    }
+
+    // Use default OpenAI-compatible format for Model APIs
+    return new OpenAICompatibleChatLanguageModel(modelId ?? 'chat', {
+      ...getCommonModelConfig('chat'),
+      errorStructure: basetenErrorStructure,
+    });
+  };
+
+  const createTextEmbeddingModel = (modelId?: BasetenEmbeddingModelId) => {
+    // Use modelURL if provided
+    const customURL = options.modelURL;
+    if (!customURL) {
+      throw new Error(
+        'No model URL provided for embeddings. Please set modelURL option for embeddings.',
+      );
+    }
+
+    // Check if this is a /sync endpoint (OpenAI-compatible) or /predict endpoint (custom)
+    // We only support /sync (not /sync/v1) to avoid double /v1 issue with Performance Client
+    const isOpenAICompatible = customURL.includes('/sync') && !customURL.includes('/sync/v1');
+
+    if (isOpenAICompatible) {
+      // Create the model using OpenAICompatibleEmbeddingModel and override doEmbed
+      const model = new OpenAICompatibleEmbeddingModel(modelId ?? 'embeddings', {
+        ...getCommonModelConfig('embedding', customURL),
+        errorStructure: basetenErrorStructure,
+      });
+
+      // Initialize the B10 Performance Client once for reuse
+      const performanceClient = new PerformanceClient(
+        customURL,
+        loadApiKey({
+          apiKey: options.apiKey,
+          environmentVariableName: 'BASETEN_API_KEY',
+          description: 'Baseten API key',
+        }),
+      );
+
+      // Override the doEmbed method to use the pre-created Performance Client
+      model.doEmbed = async params => {
```

Comment:
could you do this by subclassing `OpenAICompatibleEmbeddingModel` instead of mutating the methods like this?